runtime/common/metering.go

```diff
@@ -217,121 +232,255 @@ func max(a, b int) int {
-	// []big.Word, where big.Word = uint
+	// NOTE: big.Int.Bits() actually returns a slice of words,
+	// []big.Word, where big.Word = uint,
+	// NOT a slice of bytes!
 	return len(v.Bits()) * bigIntWordSize
 }
 
+// big.Int memory metering:
+// - |x| is len(x.Bits()), which is the length in words
+//
+
 func NewPlusBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// max(|a|, |b|) + 5
+
+	maxWordLength := max(
+		len(a.Bits()),
+		len(b.Bits()),
+	)
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		) + bigIntWordSize,
+		(maxWordLength + 5) *
+			bigIntWordSize,
 	)
 }
 
 func NewMinusBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// max(|a|, |b|) + 4
+
+	maxWordLength := max(
+		len(a.Bits()),
+		len(b.Bits()),
+	)
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
+		(maxWordLength + 4) *
+			bigIntWordSize,
 	)
 }
 
 func NewMulBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
-	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		BigIntByteLength(a) +
-			BigIntByteLength(b),
+	// if min(|a|, |b|) <= 40:
+	//     |a| + |b| + 4
+	// else:
+	//     n = min(|a|, |b|)
+	//     3 * n + max(6 * n, |a| + |b|) + 8
+
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+	minWordLength := min(
+		aWordLength,
+		bWordLength,
 	)
-}
+	wordLengthSum := aWordLength + bWordLength
 
-func NewTypeMemoryUsage(staticTypeAsString string) MemoryUsage {
-	return MemoryUsage{
-		Kind:   MemoryKindTypeValue,
-		Amount: uint64(len(staticTypeAsString)),
+	var resultWordLength int
+	if minWordLength <= 40 {
+		resultWordLength = wordLengthSum + 4
+	} else {
+		resultWordLength = 3*minWordLength + max(6*minWordLength, wordLengthSum) + 8
 	}
+	return NewBigIntMemoryUsage(
+		resultWordLength * bigIntWordSize,
+	)
 }
 
-func NewCharacterMemoryUsage(length int) MemoryUsage {
-	return MemoryUsage{
-		Kind:   MemoryKindCharacter,
-		Amount: uint64(length),
-	}
-}
+var bigOne = big.NewInt(1)
+var bigOneHundred = big.NewInt(100)
 
 func NewModBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// if a < b or |b| == 1:
+	//     |a| + 4
+	// else if |b| < 100:
+	//     |a| - |b| + 5
+	// else:
+	//     recursion_cost = pointer_size + 9 * |b| + floor(|a| / |b|) + 12
+	//     recursion_depth = 2 * BitLen(b)
+	//     3 * |b| + 4 + recursion_cost * recursion_depth
+
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+
+	var resultWordLength int
+	if a.Cmp(b) < 0 || b.Cmp(bigOne) == 0 {
+		resultWordLength = aWordLength + 4
+	} else if b.Cmp(bigOneHundred) < 0 {
+		resultWordLength = aWordLength - bWordLength + 5
+	} else {
+		recursionCost := int(unsafe.Sizeof(uintptr(0))) +
+			9*bWordLength +
+			int(math.Floor(float64(aWordLength)/float64(bWordLength))) + 12
```

Comment:
On the document, I wrote `Floor` because that's the math specific operation. In software languages, integer division implicitly returns that floor. 
We could replace `int(math.Floor(float64(aWordLength)/float64(bWordLength)))` by `aWordLength/bWordLength`.